### ISSUE TITLE
3-point calibration example not working

### ISSUE THREAD

**Issue by M1S2** [median-dispersion/XPT2046-Driver#2]
It seems that the last commit ([Refactored 3-point calibration example](https://github.com/median-dispersion/XPT2046-Driver/commit/6850790c91103874b32bd0cac2b40b5a8f986ea3)) broke the 3-point calibration example. The calibration factors are quite huge and the touch position is always at the maximum screen boundaries.

**Comment by median-dispersion** [comment]
Yeah, I updated the example while working on a bigger rewrite of the core library that might have broken compatibility with the old version. The current `XPT2046.cpp` file contains a mistake when applying the affine transformation matrix coefficients, this might be causing that behavior. I haven’t released anything yet because the rewrite will change a bunch of stuff around. I also haven’t gotten around to push those changes before going on winter holidays (sorry). The rewrite should allow for more stable touch readings by setting a minimum pressure threshold, and also allow a single calibration matrix to work with all screen rotations. So expect an initial release in a couple of weeks? (No promises though 😅...)

Regarding your requests:

1. #1
Yes, there will be a release for the Arduino library manager once the rewrite is finished. This was intended as an Arduino library first, but I will look into publishing a release for PlatformIO.

2. #2
Will be fixed with the rewrite of the core library.

3. #3
I don’t think it's technically necessary to send a power down, but the touch controller documentation is a little unclear. I had some issues with bits on the SPI bus that shouldn't be there when not powering down the touch controller. But yes, I will add an option to not send the power down command. (The rewrite will cause an increase in SPI communication though because in addition to reading the X and Y position it will also read a Z1 and Z2 position to determine the touch pressure for each sample, FYI.)

**Comment by M1S2** [comment]
Thank you for your fast response. At the moment everything is working by using a copy of the XPT2046.cpp and .h files.
But it will be good of course to use a more "official" release.
Thanks in advance and enjoy your winter holidays.

**Comment by median-dispersion** [comment]
I've had some trouble recreating the issue you were describing when using the repository at this commit: 6850790c91103874b32bd0cac2b40b5a8f986ea3. I've used PlatformIO with the Arduino framework, an ESP32-S3 and the espressif32 platform to compile the example and had no problems with calibration factors being too big or not working. Since then, I pushed the refactored library and changed some implicit casts to be explicit (fixing stricter type-checking in PlatformIO). Could you please retry using the refactored library and 3-point calibration from the [1.0.1](https://github.com/median-dispersion/XPT2046-Driver/releases/tag/1.0.1) release? If the issue persists, one thing I could imagine it being is that the compiler for the platform / board you are using is doing something different, messing with the new 3-point calibration example.

**Comment by M1S2** [comment]
Thank you for your reply and adaptions to the library.
The problem with the 3-point-calibration example unfortunately still exists with the release 1.0.1 (commit id 3dae174).
I get the following calibration matrix after doing the calibration:

```
Calibration matrix coefficients
===============================

A: 1.4800463
B: 0.2257407
C: 1.1868056
D: 2.2522223
E: 0.4503704
F: 1.8203703
W: 320
H: 240
R: 1

Use this line to calibrate the touchscreen in your sketch:

XPT2046::Calibration calibration = {1.4800463,0.2257407,1.1868056,2.2522223,0.4503704,1.8203703,320,240,1};
```
This is leading to always the maximum coordinates for X and Y (X = 320, Y = 240).

I'm using an Arduino Mega 2560 and the platformio extension in the Visual Studio Code IDE.

But I'm fine to use the old calibration example code to get a valid calibration matrix.

**Comment by M1S2** [comment]
Also a touch isn't detected as good as before (only using the IRQ pin).

**Comment by median-dispersion** [comment]
Ok that's weird... The math hasn't really changed in between revisions, the only thing that has changed is that I "un-spaghettied" the code a little bit. There are a couple of possible causes I could imagine with the new version:

1. The positions of the touch targets have changed, now they are based on a relative percentage of the screen width and height. Maybe the old positions were better for your screen? But as far as I can see as long as the determinant is not 0 for a given set of targets, causing a division by 0 error, the calculations should result in a valid calibration matrix.

2. Data types were changed throughout the code. This might cause some big errors due to narrowing conversion issues. For example: the coefficient calculation was using all `float` before and now is using the `uint16_t` values directly without explicitly converting to `float`. But the compiler should automatically choose the correct type to perform the calculation with by implicitly converting everything to `float`. And for me, it does, I get the expected result. This might be a difference between the espressif and Arduino compilers?

3. Some displays have their X or Y axis flipped. This is due to the manufacturing process where the resistive touch layer was flipped before adhering it to the display, causing inverted raw X or Y readings. But in my testing this doesn't cause the calibration to be wrong and if the old example works for you this is definitely not it!

I will try using the Arduino compiler with PlatformIO next, but I don't have an Arduino Mega 2560. I should have a couple of Arduino Nano clones flying around somewhere and will test with them. I really want to get this fixed, but **I might require your help again,** if that is ok with you? I will change some code around, add some debug info and get back to you if I can't recreate the issue by my self.

Regaring this:

> Also a touch isn't detected as good as before (only using the IRQ pin).

The responsiveness can be easily adjusted by using `touch.setTouchPressure(3.5);` and setting the threshold to a higher value, like `7.0`. You need to fine tune this for your touchscreen. I just figured that `3.5` seems fine for the display I tested with. To completely disregard touch pressure, you could set it to something like `+150.0`. I added this because on larger screens (+3.2 inch) if I didn't apply enough pressure the controller struggled to get a good measurement from the large restive layer and the touch position never settle down to a single point, making the reading really unstable and sometimes completely wrong. I would recommend a value that requires some touch pressure for more accurate and consistent readings, but on a smaller screen this might not be as much of a problem.

**Comment by M1S2** [comment]
I used the `touch.setTouchPressure(10.0)` to raise the pressure value. But the measured pressure was below 2.0 and should be ok.
But I sometimes had to touch a target over 10 times before it was recognized.

**Comment by M1S2** [comment]
It's ok, If you need some input. I will try to help, if I can.

**Comment by median-dispersion** [comment]
TLDR: This took me a **wile** to figure out, but after swapping to an Arduino Nano, I was able to replicate the issues you were describing and fix them.

> It seems that the last commit ([Refactored 3-point calibration example](https://github.com/median-dispersion/XPT2046-Driver/commit/6850790c91103874b32bd0cac2b40b5a8f986ea3)) broke the 3-point calibration example. The calibration factors are quite huge and the touch position is always at the maximum screen boundaries.

This was a type conversion issue and was fixed by explicitly converting all variables to floats before performing the calculation.

> I used the `touch.setTouchPressure(10.0)` to raise the pressure value. But the measured pressure was below 2.0 and should be ok. But I sometimes had to touch a target over 10 times before it was recognized.

This was an issue with the IRQ pin being reset and going high when transmitting data via SPI, resulting in a false "not touching" reading. This was fixed by skipping the IRQ pin check when already transmitting SPI data. You should be able to reset the touch pressure to a reasonable level.

Please try the latest [1.0.2](https://github.com/median-dispersion/XPT2046-Driver/releases/tag/1.0.2) release and see if the issues are resolved.

**Comment by M1S2** [comment]
I've tested the latest release (1.0.2) and everything is working now. Thank's for your fast support.
